### DISCUSSION TITLE
can i use command lines arguments in my apllication?

### DISCUSSION THREAD

**Discussion by AdpAtHome** [AvaloniaUI/Avalonia#13353] [Q&A]
Hello,

I''m making an gui apllication for the raspberry pi, it wil autostart with a systemd service to autostart on boot and everything is working. Now in the apllication i want to use some command line arguments when the user presses a button by example play a sound with this command

aplay /home/pi/test/test.wav 

or remove a file 

rm /home/pi/test/test.txt

or wathever command i want to use. I made an usercontrol with a textbox for the command line and a button to start the process, and i also want to get the output. This is what i tried so far but it doesn't work

```
private void PublishCommand()
{
    var proc = new Process
    {
        StartInfo = new ProcessStartInfo
        {
            FileName = "aplay",
            Arguments = "/home/pi/test/test.wav",
            RedirectStandardOutput = true,
        }
    };

    try
    {
        proc.Start();
        while (!proc.StandardOutput.EndOfStream)
        {
            string line = proc.StandardOutput.ReadLine();
            OutputTxt += line + Environment.NewLine;
        }
    }
    catch (Exception ex)
    {
        DisplayError(ex.Message);
    }
}
```

**Comment by thevortexcloud** [comment]
This works for me
`var proc = Process.Start("aplay","file.txt");`

Can you elaborate on what is not working?

Also if you really need to play sounds there are better ways to do it. Albeit they are much more complicated ways. EG Using SDL2, which has f[airly decent bindings for C#](https://github.com/flibitijibibo/SDL2-CS) and will work on Windows and Linux.

**Reply by AdpAtHome** [reply]
To play a sound was only an example, I want to play with different kind of arguments just like a terminal inside my application, if this is possible?

**Reply by thevortexcloud** [reply]
Sure, you can pass pretty much any argument to the command. A _really_ simple example would be something like

```csharp
public void PlaySound(string filePath) {
Process.Start("aplay", filePath);
}
```

If you want to embed an actual terminal. That's way more complicated. It can be done though. But you may run into problems with that. You could also write your own terminal emulator/control but that's also fairly complicated and probably another whole project.

**Reply by AdpAtHome** [reply]
Thank you i shall try when i am at home, and also i'm interested in playing a simple wave sound ( a beep)  file, i looked at sdl2 link but cant find any documentation.

**Reply by thevortexcloud** [reply]
SDL2 has fairly extensive documentation. However it's a C API. As such most of the examples are in C and the examples don't always directly translate to C#. But it's generally fairly straightforward to use the library I linked and port the logic across (all the function names and what not are exactly the same). EG here is an (frankly overcomplicated) example of using mixer to play a sound in C.

https://lazyfoo.net/tutorials/SDL/21_sound_effects_and_music/index.php

But since the API is written in C there are more steps involved than you would probably need to do if it was a standard C# library. EG you need to make sure the native library (.so file on Linux) is available for the program to use. Some things also may also return native pointers instead of objects which may need to be manually cleaned up when you don't need them anymore.

If you have never done interop with native code before in C# it can be a fairly steep learning curve. My advice is don't do it if you are not comfortable with doing it.

**Reply by thevortexcloud** [reply]
Oh also, if you just want to beep and the machine has one of those old PC speakers on it (the Pi does not have one, and neither do a lot of modern computers), you can actually use the `beep` command instead. 

https://unix.stackexchange.com/questions/1974/how-do-i-make-my-pc-speaker-beep

You can also wire a small speaker up to the Pi and just turn it on via the GPIO pins.

**Reply by AdpAtHome** [reply]
Hello @thevortexcloud  the command line argument problem is working now, the problem was with my autostart service in systemd i forgot to pass the user and the working directory, only the 'aplay' command is not working. I tried the sdl2 but i do't understand c++. do you maby have another verry simpel option? i only want to play some simpel sounds in my apllication i have now 3 .wav files but i also have them in mp3. It has to be verry simple.

like on the Windows side i use 

```
var player = new SoundPlayer(($"./Assets/Sounds/{sound}"));
player.Load();
player.Play();
```

but this doesnt work under linux

**Reply by thevortexcloud** [reply]
Sound on Linux is complicated. I would guess the problem is you don't have permission to talk to the sound server with how you are running your app. You can _probably_ fix that by running the server under a different user or changing who is allowed to connect to it.

>but this doesnt work under linux

Like many things with Linux and .NET, Microsoft does not really care about anything that does not involve a server on Linux. So they just never ported APIs like that. Your only options is to make your own sound player using a native library (like with SDL2 or OpenAL), or use an external program. If you want to go with the latter option (which is far simpler but has lots of other problems) then instead of using aplay, use a real media player. Such as MPV or VLC. That will handle far more file formats and give you some more options. EG This seems to open VLC without any type of GUI 

`cvlc FILEPATH`

VLC does also have official C# bindings if you want to give that a try (there is even an Avalonia integration if you want to show a video), which may work better for what you are doing.

https://github.com/videolan/libvlcsharp

VLC is also crossplatform so you should be able to share code between Linux and Windows.

**Comment by thevortexcloud** [comment]
Also, if you want to delete a file, there is no reason to being using the `rm` command. Just use the standard .NET file APIs. EG `File.Delete()`. It's cross platform and is built into .NET, meaning you don't need to fiddle with command arguments and will get proper exceptions out of the box if it fails.

**Reply by AdpAtHome** [reply]
yess sorry that was a bad example from me!
